### PR TITLE
Romanchuk lukashov/notifications 2

### PR DESCRIPTION
**Pull request by Sandaljero**
New notifications for applications were added

### CODE REVIEW

File: OutOfSchool/OutOfSchool.WebApi/Services/ApplicationService.cs [hunk 1]
```diff
@@ -313,13 +330,36 @@ public async Task<ApplicationDto> Update(ApplicationDto applicationDto)
             }
         }
 
-        public async Task<IEnumerable<User>> GetNotificationsRecipients(NotificationAction action, Guid objectId)
+        public async Task<IEnumerable<string>> GetNotificationsRecipients(NotificationAction action, Dictionary<string, string> additionalData, Guid objectId)
         {
-            var result = new List<User>();
+            var result = new List<string>();
+            var applications = await applicationRepository.GetByFilter(a => a.Id == objectId, "Workshop.Provider.User").ConfigureAwait(false);
+            var application = applications.FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (Sandaljero) [reply OutOfSchool/OutOfSchool.WebApi/Services/ApplicationService.cs:338]: Done

File: OutOfSchool/OutOfSchool.WebApi/Services/NotificationService.cs [hunk 2]
```diff
@@ -82,29 +81,29 @@ public async Task Create(NotificationType type, NotificationAction action, Guid
 
             var notification = new Notification()
             {
-                Id = Guid.NewGuid(),
                 Type = type,
                 Action = action,
                 CreatedDateTime = DateTimeOffset.UtcNow,
                 ObjectId = objectId,
-                Text = string.Empty,
+                Data = additionalData is null ? new Dictionary<string, string>() : additionalData,
```
[the remark this answers is not in the document]
Comment (Sandaljero) [reply OutOfSchool/OutOfSchool.WebApi/Services/NotificationService.cs]: Done

File: OutOfSchool/OutOfSchool.WebApi/Services/INotificationReciever.cs [hunk 3]
```diff
@@ -16,8 +15,9 @@ public interface INotificationReciever
         /// Get user's ids for notification.
         /// </summary>
         /// <param name="action">Action.</param>
+        /// <param name="additionalData">Additional data.</param>
         /// <param name="objectId">ObjectId.</param>
         /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
-        Task<IEnumerable<User>> GetNotificationsRecipients(NotificationAction action, Guid objectId);
+        Task<IEnumerable<string>> GetNotificationsRecipients(NotificationAction action, Dictionary<string, string> additionalData, Guid objectId);
```
[the remark this answers is not in the document]
Comment (alrom) [reply OutOfSchool/OutOfSchool.WebApi/Services/INotificationReciever.cs]: Done

File: OutOfSchool/OutOfSchool.WebApi/Services/NotificationService.cs [hunk 4]
```diff
@@ -82,29 +81,29 @@ public async Task Create(NotificationType type, NotificationAction action, Guid
...
             };
 
-            var recipients = await service.GetNotificationsRecipients(action, objectId).ConfigureAwait(false);
+            var recipients = await service.GetNotificationsRecipients(action, additionalData, objectId).ConfigureAwait(false);
 
             foreach (var user in recipients)
             {
-                notification.UserId = user.Id;
+                notification.Id = Guid.NewGuid();
+                notification.UserId = user;
```
[the remark this answers is not in the document]
Comment (alrom) [reply OutOfSchool/OutOfSchool.WebApi/Services/NotificationService.cs]: Done

File: OutOfSchool/OutOfSchool.WebApi/Services/ProviderAdminService.cs [hunk 5]
```diff
@@ -313,10 +313,41 @@ public async Task<IEnumerable<ProviderAdminDto>> GetRelatedProviderAdmins(string
 
                     dtos.Add(dto);
                 }
-
             }
 
             return dtos;
         }
+
+        public async Task FillWithProviderAdmins(Guid workshopId, List<string> users)
```
[the remark this answers is not in the document]
Comment (alrom) [reply OutOfSchool/OutOfSchool.WebApi/Services/ProviderAdminService.cs]: Done

**Review by VadymLevkovskyi — APPROVED** [review]

File: OutOfSchool/OutOfSchool.DataAccess/Models/Notification.cs [hunk 6]
```diff
@@ -11,7 +12,7 @@ public class Notification : IKeyedEntity<Guid>
         [Required]
         public string UserId { get; set; }
 
-        public string Text { get; set; }
+        public Dictionary<string, string> Data { get; set; }
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.DataAccess/Models/Notification.cs:15]: just a thought: Maybe make the column nullable? Empty dictionary is serialised as empty JSON object `{}` if i remember correctly, taking 8 bytes of space even when the data is empty. null in InnoDB takes way less :)

**Review by DmyMi — APPROVED** [review]

File: OutOfSchool/OutOfSchool.DataAccess/Models/Notification.cs (hunk 6 above)
Comment (alrom) [reply OutOfSchool/OutOfSchool.DataAccess/Models/Notification.cs:15]: Done

**Pull request merged**